File: Algoritmos_e_Estrutura_de_Dados_II/Trabalho_pratico_3/Q07.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

#define TAM 60
#define TamFila 6

struct Jogador{
    int id;
    char nome[TAM];
    int altura;
    int peso;
    char universidade[TAM];
    int anoNascimento;
    char cidadeNascimento[TAM];
    char estadoNascimento[TAM];
    struct Jogador *prox;
};
typedef struct Jogador Jogador;

struct listaJogador{
    Jogador *primeiro;
    Jogador *ultimo;
    int tam;
};
typedef struct listaJogador listaJogador;

listaJogador newLista(){
    listaJogador lista;
    lista.primeiro = lista.ultimo = (Jogador*) malloc(sizeof(Jogador));
    lista.tam = 0;
    
    return lista;
}

//funções set
void setId(int id, Jogador* jogadorLido){
    jogadorLido->id = id;
}
void setNome(char nome[], Jogador* jogadorLido){
    if(nome[0]=='\0'){
        strcpy(jogadorLido->nome, "nao informado");
    } else {
        strcpy(jogadorLido->nome, nome);
    }
}
void setAltura(int altura, Jogador* jogadorLido){
    jogadorLido->altura = altura;
}
void setPeso(int peso, Jogador* jogadorLido){
    jogadorLido->peso = peso; 
}
void setUniversidade(char universidade[], Jogador* jogadorLido){
    if(universidade[0] == '\0'){
           strcpy(jogadorLido->universidade, "nao informado");
    } else {
           strcpy(jogadorLido->universidade, universidade);
    }
}
void setAnoNascimento(int anoNascimento, Jogador* jogadorLido){
    jogadorLido->anoNascimento = anoNascimento;
}
void setCidadeNascimento(char cidadeNascimento[], Jogador* jogadorLido){
    if(cidadeNascimento[0] == '\0'){
        strcpy(jogadorLido->cidadeNascimento, "nao informado");
    } else {
        strcpy(jogadorLido->cidadeNascimento, cidadeNascimento);
    }
}
void setEstadoNascimento(char estadoNascimento[], Jogador* jogadorLido){
    if(estadoNascimento[0] == '\0'){
        strcpy(jogadorLido->estadoNascimento, "nao informado");
    } else {
        strcpy(jogadorLido->estadoNascimento, estadoNascimento);
    }
}
void setAll(int id, char nome[], int altura, int peso, char universidade[], int anoNascimento, char cidadeNascimento[], char estadoNascimento[], Jogador* jogadorLido){
    setId(id, jogadorLido);
    setNome(nome, jogadorLido);
    setAltura(altura, jogadorLido);
    setPeso(peso, jogadorLido);
    setUniversidade(universidade, jogadorLido);
    setAnoNascimento(anoNascimento, jogadorLido);
    setCidadeNascimento(cidadeNascimento, jogadorLido);
    setEstadoNascimento(estadoNascimento, jogadorLido);
}

//funcao clone
Jogador* clone(Jogador *j){
    Jogador* jClone = (Jogador*) malloc(sizeof(Jogador));
    setAll(j->id, j->nome, j->altura, j->peso, j->universidade, j->anoNascimento, j->cidadeNascimento, j->estadoNascimento, jClone);
    return jClone;
}

//funcoes split
void splitJogador (char valLido[], char valFinal[], int virgulasPuladas){
    char aux;
    int posL = 0, posF = 0;
    while(virgulasPuladas>0){
        if(valLido[posL++]==',')
            virgulasPuladas--;
    }

    aux = valLido[posL++];
    while( aux != ',' && aux != '\n' ) {
        valFinal[posF++] = aux;
        aux = valLido[posL++];
    }
    valFinal[posF] = '\0';
}


//funcao de leitura
Jogador* lerJogador(int id){
    Jogador* jogador = NULL;

    FILE* arq = fopen("/tmp/players.csv", "r");
    if(arq == NULL)
        printf("Arquivo nao aberto");
    else{
        char idChar[TAM];
        char nome[TAM];
        char altura[TAM];
        char peso[TAM];
        char universidade[TAM];
        char nascimento[TAM];
        char cidade[TAM];
        char estado[TAM];
        char line[8*TAM];

        //pular primeira linha
        while(fgetc(arq) != '\n');
        do{
            int tam = 0;
            char c;
            do{
                c = fgetc(arq);
                line[tam++] = c;
            }while ( c != '\n'  && c != EOF);
            splitJogador(line, idChar, 0);
        } while (atoi(idChar) != id && line[0] != EOF);
        
        if(line[0] == EOF){
            printf("Id inexistente");
        } else {
            jogador = (Jogador*) malloc(sizeof(Jogador));

            setId(atoi(idChar), jogador);

            splitJogador(line, nome, 1);
            setNome(nome, jogador);

            splitJogador(line, altura, 2);
            setAltura(atoi(altura), jogador);

            splitJogador(line, peso, 3);
            setPeso(atoi(peso), jogador);

            splitJogador(line, universidade, 4);
            setUniversidade(universidade, jogador);

            splitJogador(line, nascimento, 5);
            setAnoNascimento(atoi(nascimento), jogador);
            
            splitJogador(line, cidade, 6);
            setCidadeNascimento(cidade, jogador);

            splitJogador(line, estado, 7);
            setEstadoNascimento(estado, jogador);
        }   
    }
    fclose(arq);
    return jogador;
}

//calculo media
float calculoMedia(listaJogador *lista){
    Jogador *i = lista->primeiro->prox;
    float soma = 0;
    for(int j=0; j<lista->tam; j++, i = i->prox){
        soma += i->altura;
    }
    return (float) soma/lista->tam;
}
//funcoes para remocao
Jogador* remover(listaJogador *lista){
    Jogador* rem = NULL;
    if(lista->tam > 0){
        rem = clone(lista->primeiro->prox);
        free(lista->primeiro->prox);
        lista->primeiro->prox = lista->primeiro->prox->prox;
        if(lista->tam == 1){
            lista->ultimo = lista->primeiro;    
        }
        lista->tam--;
    } else {
        puts("Erro - lista vazia\n");
    }
    return rem;
}

//funcoes para insercao
void inserir(Jogador *add, listaJogador *lista){
    if (add != NULL){
        if(lista->tam == 5){
            remover(lista);
        }
        lista->ultimo->prox = add;
        lista->ultimo = lista->ultimo->prox;
        lista->tam++;
        printf("%.0f\n", calculoMedia(lista));
    } else {
        puts("Erro - jogador inexistente\n");
    }
 }

//imprimir na tela
void printJogador(Jogador* jogador, int pos){
    printf("[%d] ## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", pos, jogador->nome, jogador->altura, jogador->peso, jogador->anoNascimento, jogador->universidade, jogador->cidadeNascimento, jogador->estadoNascimento);

}

void menuFila(char in, listaJogador* lista){
    int param1;
    if(in == 'I'){
        scanf("%d", &param1);
        inserir(lerJogador(param1), lista);
    } else if(in == 'R'){
        Jogador* rem = remover(lista);
        if(rem != NULL){
            printf("(R) %s\n", rem->nome);
        }
    }
}

int main() {
    char in[20]; 
    int qtdeOperacoes;
    listaJogador lista = newLista();

    scanf("%s", in);
    while(strcmp(in, "FIM")){
        inserir(lerJogador(atoi(in)), &lista);
        scanf("%s", in);
    }
    
    scanf("%d", &qtdeOperacoes);
    while(qtdeOperacoes-- > 0){
        scanf("%s", in);
        menuFila(in[0], &lista);
    }
    
    Jogador* i = lista.primeiro->prox;
    for(int j=0; j < lista.tam; j++, i = i->prox){
        printJogador(i, j);
    }

    return 0;
}

File: Algoritmos_e_Estrutura_de_Dados_II/Trabalho_pratico_2/Q07.java
import java.util.Date;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.FileNotFoundException;
import java.io.BufferedReader;
import java.io.FileReader;

class Jogador{
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;
    
    Jogador(){
        setAll(0, "", 0, 0, "", 0, "", "");
    }

    Jogador(int id){
        ler(id);
    }
    Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        setAll(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento);
    }

    //metodos set
    public void setId(int id){
        this.id = id;
    }
    public void setNome(String nome){
        if(nome == null){
             this.nome = "nao informado";
         } else {
             this.nome = nome;
         }
    }
    public void setAltura(int altura){
        this.altura = altura;
    }
    public void setPeso(int peso){
        this.peso = peso;
    }
    public void setUniversidade(String universidade){
       if(universidade == ""){
            this.universidade = "nao informado";
        } else {
            this.universidade = universidade;
        }
    }
    
    public void setAnoNascimento(int ano){
        this.anoNascimento = ano;
    }
    public void setCidadeNascimento(String cidade){
        if(cidade == ""){
            this.cidadeNascimento = "nao informado";
        } else {
            this.cidadeNascimento = cidade;
        }
    }
    public void setEstadoNascimento(String estado){
        if(estado == ""){
            this.estadoNascimento = "nao informado";
        } else {
            this.estadoNascimento = estado;
        }
    }

    public void setAll(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        setId(id);
        setNome(nome);
        setAltura(altura);
        setPeso(peso);
        setUniversidade(universidade);
        setAnoNascimento(anoNascimento);
        setCidadeNascimento(cidadeNascimento);
        setEstadoNascimento(estadoNascimento);

    }

    //metodos get
    public int getId(){
        return id;
    }
    public String getNome(){
        return nome;
    }
    public int getAltura(){
        return altura;
    }
    public int getPeso(){
        return peso;
    }
    public String getUniversidade(){
        return universidade;
    }
    public int getAnoNascimento(){
        return anoNascimento;
    }
    public String getCidadeNascimento(){
        return cidadeNascimento;
    }
    public String getEstadoNascimento(){
        return estadoNascimento;
    }

    //split para a leitura
    private String[] split(String linha){
        String[] splitted = {"", "", "0", "0", "", "0", "", ""};
        int pos = 0;
        if(linha != null){
            for(int i=0; i<linha.length(); i++){
                String aux = "";
 
                while(i<linha.length() && linha.charAt(i) != ','){
                    aux += linha.charAt(i++);
                }
 
                splitted[pos] = aux;
                pos++;
            }
        } 
        return splitted;
    }
    //metodo de leitura
    public void ler(int id) throws NumberFormatException{
        try{
            File file = new File("/tmp/players.csv");
            BufferedReader arq = new BufferedReader( new FileReader(file) );
            arq.readLine(); //pular primeira linha das informacoes das colunas
            String[] splitted = split(arq.readLine());
            while( Integer.valueOf(splitted[0]) != id )
                splitted = split(arq.readLine());
            
            String nome = splitted[1];
            int altura = Integer.valueOf(splitted[2]);
            int peso = Integer.valueOf(splitted[3]);
            String universidade = splitted[4];
            int anoNascimento = Integer.valueOf(splitted[5]);
            String cidadeNascimento = splitted[6];
            String estadoNascimento = splitted[7];

            setAll(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento); 
        } catch(IOException e){
            MyIO.println("Um erro ocorreu.");
            e.printStackTrace();
        }
    }
    
    //metodo de escrita
    public void imprimir(){
        MyIO.println("["+ getId() + " ## " + getNome() + " ## " + getAltura() +  " ## " + getPeso() + " ## " + getAnoNascimento() + " ## " + getUniversidade() + " ## " + getCidadeNascimento() + " ## " + getEstadoNascimento() + "]");
    }

    //metodo de clone
 
    public Jogador clone() {
        return new Jogador(id, nome, altura, peso, universidade, anoNascimento, cidadeNascimento, estadoNascimento);
    }
}



class Q07{
    private static int comp = 0, troca = 0;
    private static Jogador[] aumentarArray(Jogador[] escolha, int tam){
        Jogador[] temp = new Jogador[tam+1];
        for(int i=0; i < (tam); i++){
            temp[i] = escolha[i];
        }
        temp[tam] = new Jogador();
        return temp;
    }

    private static Jogador[] insercao(Jogador[] jogador, int qtdeJogador){
        for(int i=1; i<qtdeJogador; i++){
            Jogador tmp = jogador[i].clone();
            int j = i-1;
            int jAno = jogador[j].getAnoNascimento();
            int tmpAno = tmp.getAnoNascimento();

            //operacoes para ano de tmp menor que de j
            while( (j>=0) && ( tmpAno < (jAno=jogador[j].getAnoNascimento()) )){  
                jogador[j+1] = jogador[j];
                j--;
                comp+=2;
                troca++;
            }
            
            //operacoes para anos iguais e nome de tmp menor que de j
            while( (j>=0) && (tmpAno == (jAno=jogador[j].getAnoNascimento()) && tmp.getNome().compareTo(jogador[j].getNome()) < 0 ) ){
                jogador[j+1] = jogador[j];
                j--;
                comp+=3;
                troca++;
            }
            jogador[j+1] = tmp;
            comp++;
            troca++;
        } 
        return jogador;
    }
    public static void main(String[] args){
        try{
            int tam = 0;
            Jogador[] aJogador = new Jogador[0];
            String id;
            String nomeLido;
    
            //ler id's e armazenar quantidade lida (tam) e Jogadores correspondentes
            id = MyIO.readLine();
            while(!id.equals("FIM")){
                aJogador = aumentarArray(aJogador, tam);
                aJogador[tam++].ler(Integer.valueOf(id));
                id = MyIO.readLine();
            }
            
            long inicio = new Date().getTime();
            aJogador = insercao(aJogador, tam);

            for(int i=0; i<tam; i++){
                aJogador[i].imprimir();
            }
            long fim = new Date().getTime();
            File arq = new File("matricula_insercao.txt");


            FileWriter fw = new FileWriter(arq);
            fw.write("805413\t" + "\t" + comp+ "\t" + troca + "\t" + (fim-inicio)/1000.0);
            fw.close();
        } catch (IOException e){
            e.printStackTrace();
        }catch(NumberFormatException e){
            MyIO.println("Id invalido");
        }

    }
}
